log/logger.go

```diff
@@ -42,34 +121,206 @@ func File(path string) (logger *Logger, loggerClose func(), err error) {
 	if err := os.MkdirAll(filepath.Dir(filePath), 0755); err != nil {
-		return nil, nil, pkgerrors.WithStack(err)
+		return nil, pkgerrors.WithStack(err)
 	}
 
 	file, err := os.OpenFile(filePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
 	if err != nil {
-		return nil, nil, pkgerrors.WithStack(err)
+		return nil, pkgerrors.WithStack(err)
 	}
-	loggerClose = func() {
-		if err := file.Close(); err != nil {
-			panic(err)
+
+	writer = io.MultiWriter(parent, file)
+
+	return writer, nil
+}
+
+// spawningHandler is a structural logging handler which spawns a new log file if one of the given log file spawners triggers.
+type spawningHandler struct {
+	// writer holds the writer to write the output.
+	writer io.Writer
+
+	// attributes holds the attributes handed to the logger.
+	attributes map[AttributeKey]string
+
+	// logFileSpawners holds the spawners responsible for spawning a new log file.
+	logFileSpawners []logFileSpawner
+
+	// flags holds the flags deciding which fields are logged.
+	flags Flags
+}
+
+// newSpawningHandler returns a new spawning handler.
+func newSpawningHandler(writer io.Writer, flags Flags) *spawningHandler {
+	return &spawningHandler{
+		writer: writer,
+
+		attributes: map[AttributeKey]string{},
+
+		flags: flags,
+	}
+}
+
+// Clone returns a copy of the object.
+func (h *spawningHandler) Clone() (clone *spawningHandler) {
+	return &spawningHandler{
+		writer: h.writer,
+
+		attributes: maps.Clone(h.attributes),
+
+		logFileSpawners: slices.Clone(h.logFileSpawners),
+	}
+}
+
+var _ slog.Handler = (*spawningHandler)(nil)
+
+// Enabled reports whether the handler handles records at the given level.
+func (h *spawningHandler) Enabled(ctx context.Context, level slog.Level) bool {
+	return true
+}
+
+// Handle handles the Record.
+func (h *spawningHandler) Handle(ctx context.Context, record slog.Record) error {
+	if h.flags&FlagDate != 0 {
+		fmt.Fprint(h.writer, record.Time.Format("2006/01/02"))
+		fmt.Fprint(h.writer, " ")
+	}
+	if h.flags&FlagTime != 0 {
+		fmt.Fprint(h.writer, record.Time.Format("15:04:05"))
+		fmt.Fprint(h.writer, " ")
+	}
+
+	fmt.Fprintln(h.writer, record.Message)
+
+	return nil
+}
+
+// WithAttrs returns a new Handler whose attributes consist of both the receiver's attributes and the arguments.
+// The Handler owns the slice: it may retain, modify or discard it.
+func (h *spawningHandler) WithAttrs(attributes []slog.Attr) slog.Handler {
+	// Collect attributes.
+	for _, attribute := range attributes {
+		h.attributes[AttributeKey(attribute.Key)] = attribute.Value.String()
+	}
+
+	newHandler := h.Clone()
+
+	// Check if we need to spawn a new log file.
+	for i, spawner := range h.logFileSpawners {
+		if spawner.NeedsSpawn(h.attributes) {
+			logFilePath := spawner.FilePath(h.attributes)
+			writer, err := newLogWriter(h.writer, logFilePath)
+			if err != nil {
+				fmt.Fprintf(h.writer, "ERROR: cannot create new handler: %s\n", err.Error())
+			}
```

Comment:
should probably return from the function then?

Reply:
Not sure. We still have a valid logger where we can print the error message to. returning a nil handler here seems  to be not intended.

Reply:
the writer will be `nil` though, right? so maybe we just reset the writer to the one we were using

Reply:
the new writer is nil, but we are logging to it's parent `h.writer`

Reply:
below we set `newHandler.writer = writer` and return that as new handler... so when that handler is used and tries to log to its own writer :boom:

Reply:
I think in this case we should do `continue` and ignore the current spawner

Reply:
ah sure. the `continue` is missing. sorry.